src/host/dns.h

```diff
@@ -46,6 +55,10 @@ namespace asynchost
             host,
             service,
             uv_strerror(rc));
+          {
+            std::unique_lock<std::mutex> guard(pending_resolve_requests_mtx);
+            pending_resolve_requests.erase(resolver);
+          }
```

Comment:
Should we only append to `pending_resolve_requests` if `uv_getaddrinfo()` is successful?

Reply:
I'm not sure about that. Is it guaranteed that the callback will not be run if `uv_getaddrinfo(...) < 0`? If so, then we can do the insert afterwards, because we won't need `resolver->data`.

Reply:
http://docs.libuv.org/en/v1.x/dns.html#api

> If successful, the callback will get called sometime in the future with the lookup result

Which I believe implies that the callback won't be called if the call isn't successful?

Reply:
Hmm. I think I may have been wrong about this. If we were to add it after a successful call to `uv_getaddrinfo`, then `~TCPImpl` and the callback could be executed (in both orders) before we add it to the set, potentially allowing an invalid pointer to slip through.